sdk/python/feast/client.py

```diff
@@ -508,36 +528,186 @@ def ingest(
         else:
             raise Exception(f"Feature set name must be provided")
 
+        time_start = time.time()
+
+        # Read table and get row count
         table = _read_table_from_source(source)
```

Comment:
Is all data always read into memory? Would it be better to read in chunks rather than read all then chunk?

Reply:
I'm not very sure if reading chunks into memory is possible within pyarrow's parquet read_table(). 

I might have missed something in the documentation. Do let me know if you're able to find something that allows this.

pyarrow.parquet.ParquetFile exposes `read_row_groups()`, this is under the assumption that the parquet file has more than 1 row groups, which is usually not the case.

Link: https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetFile.html#pyarrow.parquet.ParquetFile.read_row_group

For a parquet file to have more than 1 row group (should it only have one row group), we can read parquet file into a table and write it out again by setting the `row_group_size` in pyarrow.parquet.write_table. 

Link:
https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_table.html#pyarrow.parquet.write_table

https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html?highlight=row_group_size

I can try implementing this, but it will still require a write to the filesystem nonetheless.

Reply:
Perhaps, the best approach one can take now is to:

1) Determine the number of row groups there are in a ParquetFile.
2) If the number of row groups is equals to one, create another table with `row_group_size` equal to the original `chunk_size`
3) Pass the ParquetFile handle or directory to the multiprocessing pool workers together with the `row_group_index` it should be working on.
4) Everything downstream will remain the same.

What do you think?

Reply:
Is it possible to use https://arrow.apache.org/docs/python/generated/pyarrow.RecordBatchFileReader.html?

Reply:
I checked, this is not possible, documentation says it's only for Arrow binary files.

Just for good measure to see if there any side effects in this class.

```
import pyarrow as pa

foo = pa.RecordBatchFileReader("parquet_file.parquet")
```

returns

```
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "../feast/venv/lib/python3.7/site-packages/pyarrow/ipc.py", line 98, in __init__
    self._open(source, footer_offset=footer_offset)
  File "pyarrow/ipc.pxi", line 396, in pyarrow.lib._RecordBatchFileReader._open
  File "pyarrow/error.pxi", line 78, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Not an Arrow file
```

Reply:
I see. Okay then, not being able to ingest files that can't fit into memory is a concern but there are workarounds and it can be considered out of scope for this PR. thanks!

Reply:
I'll do the changes to `read_num_groups()` of a ParquetFile object instead of splitting a large PyArrow Table into smaller PyArrow Table and writing it to parquet files.

Although this will not solve the issue of not being able to ingest files that cannot fit into memory, this should cleanup the implementation a little. i.e. Instead of creating multiple parquet files in the file system, a parquet files with multiple row groups is created instead.

Reply:
@zhilingc I've made the changes for using `read_num_groups()` instead of splitting the PyArrow Table into multiple individual parquet files. 

The row groups are now distributed to pool workers instead of file directory references.